owid/owid-grapher
# Lerna

Adds lerna to the top of the repo to manage three new sub packages:
- `@ourworldindata/utils`
- `@ourworldindata/core-table`
- `@ourworldindata/grapher`

There's also an empty `@ourworldindata/types` package which is unused at the moment but I intend to add to later.

Lerna works best when every project is in a workspace it's watching, but for now we still have the site, client, server, etc outside of Lerna's scope. 

This means that our top level `package.json` still has the dependencies for all these projects and they all have to be installed with `yarn install` before running `npx lerna bootstrap`

To test, all you should need to do is:
1. `rm -rf node_modules`
2. `make up.full`

The only business logic change to the project is that we now no longer bake Grapher with its URL dependencies hard-coded, they must be [passed via config during instantiation instead](https://github.com/owid/owid-grapher/commit/df315c62da150d90a084efe39da6ee5e695f327e). 

`utils` and `core-table` both build with `tsc` to output a declaration and js files in their `dist/` directories. `grapher` uses [tsup](https://tsup.egoist.dev) which is an esbuild wrapper in order to also bundle the CSS.

esbuild runs extremely quickly if you opt to skip type checking. ~I still need to implement a `watch` script for local grapher development that will try to leverage this.~
